### ISSUE TITLE
net/http: http.FileServer goes into infinite redirect loop when root is a file

### ISSUE THREAD

**Issue by meblum** [golang/go#63769]
### What version of Go are you using (`go version`)?

<pre>
go1.21.3 windows/amd64
</pre>

### Does this issue reproduce with the latest release?
Yes

### What operating system and processor architecture are you using (`go env`)?

<details><summary><code>go env</code> Output</summary><br><pre>
set GO111MODULE=
set GOARCH=amd64
set GOBIN=
set GOCACHE=C:\Users\MeirBloomenfeld\AppData\Local\go-build
set GOENV=C:\Users\MeirBloomenfeld\AppData\Roaming\go\env
set GOEXE=.exe
set GOEXPERIMENT=
set GOFLAGS=
set GOHOSTARCH=amd64
set GOHOSTOS=windows
set GOINSECURE=
set GOMODCACHE=C:\Users\MeirBloomenfeld\go\pkg\mod
set GONOPROXY=
set GONOSUMDB=
set GOOS=windows
set GOPATH=C:\Users\MeirBloomenfeld\go
set GOPRIVATE=
set GOPROXY=https://proxy.golang.org,direct
set GOROOT=C:\Program Files\Go
set GOSUMDB=sum.golang.org
set GOTMPDIR=
set GOTOOLCHAIN=auto
set GOTOOLDIR=C:\Program Files\Go\pkg\tool\windows_amd64
set GOVCS=
set GOVERSION=go1.21.3
set GCCGO=gccgo
set GOAMD64=v1
set AR=ar
set CC=gcc
set CXX=g++
set CGO_ENABLED=0
set GOMOD=C:\Users\MeirBloomenfeld\Downloads\test\go.mod
set GOWORK=
set CGO_CFLAGS=-O2 -g
set CGO_CPPFLAGS=
set CGO_CXXFLAGS=-O2 -g
set CGO_FFLAGS=-O2 -g
set CGO_LDFLAGS=-O2 -g
set PKG_CONFIG=pkg-config
set GOGCCFLAGS=-m64 -fno-caret-diagnostics -Qunused-arguments -Wl,--no-gc-sections -fmessage-length=0 -ffile-prefix-map=C:\Users\MEIRBL~1\AppData\Local\Temp\go-build4162283884=/tmp/go-build -gno-record-gcc-switches

</pre></details>

### What did you do?

```go
func main() {
	// myFile is a file, not a directory
	http.ListenAndServe(":8080", http.FileServer(http.Dir("myFile")))
}
```

### What did you expect to see?
a request to http://localhost:8080 returns an error response, or the content of "myFile"

### What did you see instead?

server responds with 301 status code to redirect to "..//" which causes an infinite redirect loop.

**Comment by seankhliao** [comment]
cc @neild

**Comment by meblum** [comment]
Currently, when root is a file, requests to the root creates redirect loop, and other paths return a proper 404. Since the root is expected to be a directory, it probably makes sense to return a 500 for all requests to a fileserver which has a file set as http.Dir.

**Comment by meblum** [comment]
Note, the current implementation of ServeFileFS will return the file when using it as root.

**Comment by mauri870** [comment]
The http.FileSystem interface predates fs.FS and might not be 100% compatible in some cases, that is why ServeFileFS works as expected, been a newer api.

Having this routine enter an infinite redirect loop is not cool, and appears to be a bug that should be addressed properly.

**Comment by mauri870** [comment]
It is unclear to me what should happen if http.Dir is invoked with a normal file as argument, and I could not find any tests or documentation around it to back it up. All the occurrences of http.Dir in the codebase use a directory as parameter. 

Is that right to assume it is undefined behavior calling `http.FileServer(http.Dir("myFile"))`?

I'll send a CL shortly that allows a normal file to be served using `http.FileServer(http.Dir("myFile"))`, but I'm not sure if that is something it should support, or if we should error out at some point saying that we need a real directory.

**Comment by meblum** [comment]
The reason why ServeFileFS doesn't have this redirect loop issue, is because like ServeFile, it doesn't redirect requests at all (index.html being the exception), it just returns the content found at path. I don't think it was a deliberate api design decision not to error when using a file instead of a directory. However we probably shouldn't be changing that now to return an error response as it will break users relying on this behavior.

But for FileServer returning an error for this case won't be a breaking change and it might even make sense to return an error instead of the file, since we expect a file (other than index.html) to be served only from a request not ending with a slash.
In other words, I believe this will be the only case of a file other than index.html being served for a path ending in a slash. Does it matter?

**Comment by mauri870** [comment]
@neild @bcmills  I'm not quite sure what should be the path forward to fix this but I would like to merge the fix before the freeze. I believe we have two choices:

1. We make Dir handle files and document that it can do that, and end up with a constructor like `Dir("file.txt")` which looks quite odd but it works and fixes the redirect loop.
2. We call Stat inside `(Dir).Open` and fail if the provided file is not a directory, and we document that as well. That should fix the redirect loop too.

I'm inclined to go with option 2, wdyt?

**Comment by bcmills** [comment]
I prefer option (2) from the perspective of doing [what it says on the tin](http://thecodelesscode.com/case/84), but I will defer to Damien if he has a preference.

**Comment by odeke-em** [comment]
Thank you @mauri870 for the CL, I've commented on your CL per https://go-review.googlesource.com/c/go/+/538719/comments/dfdfd7d9_f04c9321 about strictly ensuring that we only perform point 2 and that we preserve the old behavior as is, following through with https://www.hyrumslaw.com/ and only stat-ing to ensure that we have purely directories.

**Comment by mauri870** [comment]
@odeke-em Thanks for the feedback, I addressed your comments in the CL.
